src/index_notation/provenance_graph.cpp

```diff
@@ -875,8 +815,25 @@ std::vector<ir::Expr> PrecomputeRelNode::deriveIterBounds(taco::IndexVar indexVa
                                                      std::map<taco::IndexVar, taco::ir::Expr> variableNames,
                                                      Iterators iterators,
                                                      ProvenanceGraph provGraph) const {
+
   taco_iassert(indexVar == getPrecomputeVar());
   taco_iassert(parentIterBounds.count(getParentVar()) == 1);
+
+  if (indexVar.isBound()){
```

Comment:
It's quite weird that each relation node's implementation now has to check things about bound. That's not a modular design

Reply:
Changed to remove this check for deriveIterBounds from the RelNodes functions to the deriveIterBounds for the provenance graph (so, we only need to make this check once).